ligo/src/management.mligo

```diff
@@ -4,36 +4,74 @@
 
 #include "types.mligo"
+#include "common.mligo"
 
-
-// -----------------------------------------------------------------
-// Shared Function
-// -----------------------------------------------------------------
 #if !MANAGEMENT_H
 #define MANAGEMENT_H
 
-[@inline]
-let ensure_not_migrated (storage : storage): storage =
-  match storage.migration_status with
-      Not_in_migration -> storage
-    | MigratingTo (p) -> storage
-    | MigratedTo (p) -> (failwith ("MIGRATED") : storage)
-
-#endif  // MANAGEMENT_H included
-
-
+(*
+ * Auth checks for admin and store the address in parameter to the
+ * 'pending_owner' field in storage.
+ *)
 let transfer_ownership
     (param, store : transfer_ownership_param * storage) : return =
-  not_implemented("transfer_ownership")
+      if authorize_admin(store)
+        then
+          (([] : operation list), { store with pending_owner = param ; })
+        else (failwith("NOT_ADMIN") : return)
 
+(*
+ * Auth check for pending admin and copies the value in 'pending_owner' field
+ * to 'admin' field. The 'pending_owner' field is left with the address of the
+ * new admin.
+ *)
 let accept_ownership(param, store : unit * storage) : return =
-  not_implemented("transfer_ownership")
+  if store.pending_owner = Tezos.sender
+    then (([] : operation list), { store with admin = store.pending_owner })
+    else (failwith("NOT_PENDING_ADMIN") : return)
 
+(*
+ * Auth check for admin and sets the migration status to 'MigratingTo' using
+ * the address from parameter, overwriting any address from any previous
+ * 'Migrate' calls.
+ *)
 let migrate(param, store : migrate_param * storage) : return =
-  not_implemented("migrate")
+  if authorize_admin(store)
+    then (([] : operation list), { store with migration_status = MigratingTo (param) })
+    else (failwith("NOT_ADMIN") : return)
 
+(*
+ * Auth check for new contract and sets the migration_status to 'Migrated', using
+ * the address from the earlier value in the migration_status field.
+ *)
 let confirm_migration(param, store : unit * storage) : return =
-  not_implemented("confirm_migration")
+  match store.migration_status with
+    Not_in_migration -> (failwith("NOT_MIGRATING") : return)
+  | MigratingTo (new_addr) -> if new_addr = Tezos.sender
+      then (([] : operation list), { store with migration_status = MigratedTo (new_addr) })
+      else (failwith("NOT_MIGRATION_TARGET") : return)
+  | MigratedTo (new_addr)  -> (failwith("MIGRATED") : return)
+
+(*
+ * Call a custom entrypoint. Looks up the packed code for entrypoint using the entrypoint
+ * name in the parameter. Then unpacks the code to a lambda of type 'bytes -> return'.
+ * Then just pass the bytes from parameter to this lambda return the resulting value.
+ *
+ * NO AUTH CHECKS ARE DONE.
+ *)
+let call_custom(param, full_storage : custom_ep_param * full_storage) : return =
+  let ep_name = param.0 in
+  let config = full_storage.1 in
+  let storage = full_storage.0 in
```

Comment:
At this implementation it seems that the `full_storage` parameter could be replaced by the two `config` and `store` one (since we already made the un-pairing in `base_DAO.mligo`), however I was actually thinking that the `lambda` could accept the `full_storage` instead of the `storage`, so this would be fine if we did that.

The reason for this change in the lambda is that since this LIGO version is being made with runtime-customization in mind (is more flexible, whereas the Lorentz one is more static but also more efficient) we could have more interesting things happen in these custom entrypoints.
One example that was being considered is to have an entrypoint that can only be run once because during that first run it removes itself from the `map` (to reduce the cost of origination).

What do you think? can we do that?
If so please also update the `Custom entrypoints` part of the specification as well to use `full_storage` instead of `storage`.

Reply:
I had the impression that we wanted to restrict the custom entrypoints from changing the configuration in the storage. I think the use case discussed in the channel makes sense. So I have changed the lambdas to be `((bytes * full_storage) -> (operation list * full_storage))`. 

One thing though, I had to refactor the top level `case` a bit to accomodate for the fact that the customCall now returns a full_storage, unlike the other calls that returns `storage` only.